### PR TITLE
Feature/instrument type time alloc

### PR DESCRIPTION
**Pull request by Fingel**
Adds instrument name to TimeAllocation, so that allocations are granular to the instrument name level. Also does time accounting on a request by request basis, instead of full accounting per proposal on each run.

### CODE REVIEW

File: valhalla/proposals/models.py [hunk 1]
```diff
@@ -103,6 +114,7 @@ class TimeAllocation(models.Model):
     semester = models.ForeignKey(Semester)
     proposal = models.ForeignKey(Proposal)
     telescope_class = models.CharField(max_length=20, choices=TELESCOPE_CLASSES)
+    instrument_name = models.CharField(max_length=200, choices=INSTRUMENT_NAMES)
```
Comment (jnation3406) [inline valhalla/proposals/models.py]: can the choices be dynamic? I.e. leave choices=() but in forms/admin that uses it get it dynamically from configdb?

Comment (jnation3406) [inline valhalla/proposals/models.py]: can we call this variable something other than instrument_name? instrument_name often means the specific instrument's name (i.e. fl06). What about camera_type, or instrument_type, or instrument_class or camera_class?

File: valhalla/proposals/tasks.py [hunk 2]
```diff
@@ -1,29 +1,39 @@
...
+            talloc = request.timeallocation
+            logger.info('Performing accounting for request: %s', request.id, extra=log_tags)
+            time_used = query_pond_for_time_used(request)
+            if request.user_request.observation_type == UserRequest.TOO:
+                talloc.too_time_used += time_used['too_time']
+            else:
+                talloc.std_time_used += time_used['std_time']
+            talloc.save()
+            request.accounted = True
+            request.save()
```
Comment (jnation3406) [inline valhalla/proposals/tasks.py]: I think this should be wrapped in @transaction.atomic like the functions that modify request state in state_changes.py are.

File: valhalla/userrequests/models.py [hunk 3]
```diff
@@ -175,6 +178,10 @@ def blocks(self):
             logger.error('Could not connect to the pond.')
             return BlockSerializer([], many=True).data
 
+    @classmethod
+    def need_accounting(clazz):
+        return clazz.objects.filter(state__in=['COMPLETED', 'CANCELED', 'WINDOW_EXPIRED'], accounted=False)
```
Comment (jnation3406) [inline valhalla/userrequests/models.py]: we have a TERMINAL_STATES list in state_changes.py, maybe we should pull that into the class and reuse it here.

File: valhalla/proposals/models.py (hunk 1 above)
Comment (Fingel) [reply valhalla/proposals/models.py]: I'm not sure how we could pull this off. Django needs a set list of values for validation. We could make it a freeform field, but we would risk serious bugs from typos.

Comment (Fingel) [reply valhalla/proposals/models.py]: It's named `instrument_name` in Molecule, so I'd prefer to keep things consistent, at least within Valhalla. I'm aware that it is inconsistent in other places in the stack.

File: valhalla/proposals/tasks.py (hunk 2 above)
Comment (Fingel) [reply valhalla/proposals/tasks.py]: Good idea.

File: valhalla/userrequests/models.py (hunk 3 above)
Comment (Fingel) [reply valhalla/userrequests/models.py]: Good idea.

File: valhalla/proposals/models.py (hunk 1 above)
Comment (jnation3406) [reply valhalla/proposals/models.py]: good point

File: valhalla/proposals/migrations/0009_auto_20171003_2017.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
```
Comment (jnation3406) [inline valhalla/proposals/migrations/0009_auto_20171003_2017.py]: why two migrations?

File: valhalla/proposals/models.py [hunk 5]
```diff
@@ -94,6 +94,17 @@ class TimeAllocation(models.Model):
...
 
+    INSTRUMENT_NAMES = (
+        ('0M4-SCICAM-SBIG', '0M4-SCICAM-SBIG'),
+        ('0M8-NRES-SCICAM', '0M8-NRES-SCICAM'),
+        ('0M8-SCICAM-SBIG', '0M8-SCICAM-SBIG'),
+        ('1M0-NRES-SCICAM', '1M0-NRES-SCICAM'),
+        ('1M0-SCICAM-SINISTRO', '1M0-SCICAM-SINISTRO'),
+        ('1M0-SCICAM-SBIG', '1M0-SCICAM-SBIG'),
+        ('2M0-FLOYDS-SCICAM', '2M0-FLOYDS-SCICAM'),
+        ('2M0-SCICAM-SPECTRAL', '2M0-SCICAM-SPECTRAL')
```
Comment (jnation3406) [inline valhalla/proposals/models.py]: I guess we should add the 1M0-NRES-COMMISSIONING instrument to the list

File: valhalla/userrequests/state_changes.py [hunk 6]
```diff
@@ -106,7 +106,7 @@ def debit_ipp_time(ur):
         return
     try:
         time_allocations = ur.timeallocations
```
Comment (jnation3406) [inline valhalla/userrequests/state_changes.py]: does line ~84, validate_ipp function need changes too?

File: valhalla/proposals/migrations/0009_auto_20171003_2017.py (hunk 4 above)
Comment (Fingel) [reply valhalla/proposals/migrations/0009_auto_20171003_2017.py]: One was to add the field, the other adds a choice. We can squash them if it bothers you.

File: valhalla/userrequests/state_changes.py (hunk 6 above)
Comment (Fingel) [reply valhalla/userrequests/state_changes.py]: Yea, it did.

Comment (Fingel) [reply valhalla/userrequests/state_changes.py]: Where are the unit tests for this function? Can't find any. I would expect them to fail.

**Pull request merged**